README.md
# Advanced configurable layered Box Shadow classes for tailwindcss

# tailwindcss-layered-box-shadow

A plugin that provides advanced, more realistic and fully configurable layered box shadow classes for tailwindcss. 

## Installation

Install the plugin from npm:

```sh
# Using npm
npm install tailwindcss-layered-box-shadow

# Using Yarn
yarn add tailwindcss-layered-box-shadow
```

Then add the plugin to your `tailwind.config.js` file:

```js
// tailwind.config.js
module.exports = {
  theme: {
    // ...
  },
  plugins: [
    require('tailwindcss-layered-box-shadow'),
    // ...
  ],
}
```

## Usage

The Plugin generates 6 default advanced shadow sizes with different number of layers:

| Class                | Layers |
|----------------------|--------|
| `shadow-layered-xs`  | `3`    |
| `shadow-layered-sm`  | `4`    |
| `shadow-layered-md`  | `5`    |
| `shadow-layered-lg`  | `6`    |
| `shadow-layered-xl`  | `7`    |
| `shadow-layered-2xl` | `8`    |

## Example Usage

```html
<div class="shadow-layered-xs hover:shadow-layered-lg"></div>
```

Output:

```css
.shadow-layered-xs {
    box-shadow: 2px 2px 10px -0.2px rgba(98,103,173,0.34),
    3px 3px 10px -0.575px rgba(98,103,173,0.34),
    4px 4px 10px -0.95px rgba(98,103,173,0.34)
;
}

hover\:.shadow-layered-lg {
    box-shadow: 2px 2px 10px -0.2px rgba(98,103,173,0.34),
    3px 3px 10px -0.575px rgba(98,103,173,0.34),
    4px 4px 10px -0.95px rgba(98,103,173,0.34),
    5px 5px 10px -1.325px rgba(98,103,173,0.34),
    6px 6px 10px -1.7px rgba(98,103,173,0.34),
    7px 7px 10px -2.075px rgba(98,103,173,0.34)
;
}
```

## Configuration

You can configure the shadow styles are generated by this plugin under the `shadowColor`, `shadowPositionX`, `shadowPositionY`, `shadowBlur`, `shadowSpread` and `shadowOpacity` keys in your `tailwind.config.js` file:

```js
// tailwind.config.js
module.exports = {
    theme: {
        shadowColor: '#6267ad',
        shadowPositionX: 1,
        shadowPositionY: 1,
        shadowBlur: 10,
        shadowSpread: -0.2,
        shadowOpacity: 0.34
    }
}
```

## Demo

[https://bentzibentz.github.io/tailwindcss-layered-box-shadow/](https://bentzibentz.github.io/tailwindcss-layered-box-shadow/)


CHANGELOG.md


### [0.0.2](https://github.com/bentzibentz/tailwindcss-layered-box-shadow/compare/0.0.1...0.0.2) (2022-09-03)

### 0.0.1 (2022-09-03)

scripts/build.js
const fs = require('fs')
const postcss = require('postcss')
const tailwind = require('tailwindcss')
const CleanCSS = require('clean-css')

function buildDistFile(filename) {
  return postcss([
    tailwind({
      corePlugins: false,
      plugins: [require('../src/index.js')],
    }),
    require('autoprefixer'),
  ])
    .process('@tailwind utilities', {
      from: null,
      to: `./dist/${filename}.css`,
      map: false,
    })
    .then((result) => {
      fs.writeFileSync(`./dist/${filename}.css`, result.css)
      return result
    })
    .then((result) => {
      const minified = new CleanCSS().minify(result.css)
      fs.writeFileSync(`./dist/${filename}.min.css`, minified.styles)
    })
    .catch((error) => {
      console.log(error)
    })
}

console.info('Building CSS...')

Promise.all([buildDistFile('layered-box-shadow')]).then(() => {
  console.log('Finished building CSS.')
})


src/index.js
const plugin = require('tailwindcss/plugin')

/**
 * ECMA2016 / ES6 Convert hex to rgb
 */
const convertHexToRGBA = (hexCode, opacity) => {
    let hex = hexCode.replace('#', '');

    if (hex.length === 3) {
        hex = `${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}`;
    }

    const r = parseInt(hex.substring(0, 2), 16);
    const g = parseInt(hex.substring(2, 4), 16);
    const b = parseInt(hex.substring(4, 6), 16);

    if (opacity > 1 && opacity <= 100) {
        opacity = opacity / 100;
    }

    return `${r},${g},${b},${opacity}`;
};

const layeredBoxShadow = plugin(
    function ({addUtilities, theme, e}) {
        const shadowColor = theme('shadowColor');
        const shadowOpacity = theme('shadowOpacity');
        const shadowPositionX = theme('shadowPositionX');
        const shadowPositionY = theme('shadowPositionY');
        const shadowBlur = theme('shadowBlur');
        const shadowSpread = theme('shadowSpread');

        const color = convertHexToRGBA(shadowColor, shadowOpacity ? shadowOpacity : 1);

        const layers = [
            {size: 'xs', layers: 3},
            {size: 'sm', layers: 4},
            {size: 'md', layers: 5},
            {size: 'lg', layers: 6},
            {size: 'xl', layers: 7},
            {size: '2xl', layers: 8}
        ];

        const generateShadows = () => {
            let shadows = [];
            layers.forEach((el) => {
                const layers = Array(el.layers).fill(0);
                let shadowValueStr = '';
                let blur = 0;
                let spread = shadowSpread;
                layers.forEach((el, idx) => {
                    if (blur === 0) {
                        blur = shadowBlur;
                    }
                    shadowValueStr = shadowValueStr + ' ' + `${shadowPositionX * (idx + 2)}px ${shadowPositionY * (idx + 2)}px ${blur}px  ${spread}px rgba(${color}),`;
                    spread = spread - 0.375;
                })

                shadowValueStr = shadowValueStr.slice(0, -1);

                const shadow = {[el.size]: shadowValueStr};

                shadows.push(shadow);
            });

            return shadows;
        }

        const shadowArrs = generateShadows();
        const shadows = Object.assign({}, ...shadowArrs);

        addUtilities(
            [
                Object.entries(shadows).map(([key, value]) => {
                    return {
                        [`.${e(`shadow-layered-${key}`)}`]: {
                            'boxShadow': `${value}`,
                        },
                    }
                })
            ]
        )
    },
    {
        experimental: {
            matchVariant: true
        },
        theme: {
            shadowColor: '#6267ad',
            shadowPositionX: 1,
            shadowPositionY: 1,
            shadowBlur: 10,
            shadowSpread: -0.2,
            shadowOpacity: 0.34
        },
    }
)

module.exports = layeredBoxShadow;
